File: test/python/shading.py
import Magics
from numpy import *

Magics.init ()

formats = ["pdf", "PNG", "ps"]

Magics.set1c("output_formats",formats)
Magics.setc("output_name","shading_py")

colour = ["NAVY", "BLUE", "CYAN", "GREEN","YELLOW", "ORANGE", "RED", "PURPLE"]
level  =  array([480., 490., 500., 530., 545., 560., 575., 590., 600.])
markers  = array([0,1, 2, 3, 4, 5, 6, 7, 8], dtype=int32)
print markers.dtype
test   = array([450., 515., 530.])


Magics.setc ("legend", "on")
Magics.setc ("map_coastline_colour","grey")
Magics.setc ("map_grid_colour","grey")   
    
# pass the data to Magics

Magics.setc ("grib_input_type","file")
Magics.setc ("grib_input_file_name","../data/z500.grb")
Magics.grib ()
      
 
# define the contour     
     
Magics.setc ("CONTOUR_LEVEL_SELECTION_TYPE", "LEVEL_LIST")
Magics.set1r("CONTOUR_LEVEL_LIST", level)
Magics.set1i("contour_shade_marker_table",markers)
Magics.setc ("CONTOUR_SHADE",            "ON")      
Magics.setc ("CONTOUR_SHADE_TECHNIQUE",  "marker")
Magics.setc ("contour_line_colour",      "evergreen")
Magics.setc ("contour_highlight_colour", "evergreen")
Magics.setc ("CONTOUR_SHADE_COLOUR_METHOD", "list")
Magics.set1c("CONTOUR_SHADE_COLOUR_LIST", colour)
Magics.setc ("contour_hilo","off")     
Magics.setc ("legend","on")  
Magics.cont ()
      

# generate text for title
Magics.setc ("text_line_1", "Contours with HATCH shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")

# plot the coastlines and title

Magics.text ()
Magics.coast ()
Magics.new_page("PAGE")
      
Magics.setc ("CONTOUR_SHADE_METHOD",     "DOT")
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    45.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -58.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   55.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE", -40.0)

Magics.setc ("CONTOUR_SHADE_TECHNIQUE",  "POLYGON_SHADING")
Magics.setc ("CONTOUR_SHADE_METHOD",     "HATCH")
Magics.cont ()
Magics.coast ()
  

Magics.setc ("text_line_1", "Contours with DOT shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text ()
      
      
Magics.new_page("PAGE")

Magics.setc ("CONTOUR_SHADE_METHOD",     "AREA_FILL")
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    40.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -56.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   55.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",   -20.0)
Magics.cont ()
Magics.coast ()
      
Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text ()
      
      
Magics.new_page("PAGE")
      
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    55.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -56.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   75.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",   -20.0)
Magics.cont ()
Magics.coast ()

Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text ()
        
Magics.new_page("PAGE")
      
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    10.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  20.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   60.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",   70.0)
Magics.cont ()
Magics.coast ()
Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text ()
      
            
Magics.new_page("PAGE")
      
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    -70.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -98.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   -30.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",   -50.0)
Magics.cont()
Magics.coast()

Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text()
               
Magics.new_page("PAGE")
Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    -40.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -118.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   -30.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",   -102.0)
Magics.cont()
Magics.coast()

Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text()
      
Magics.new_page("PAGE")

Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    30.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  -30.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   50.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",  70.0)
Magics.cont()
Magics.coast()

Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text()

Magics.new_page("PAGE")

Magics.setr ("SUBPAGE_LOWER_LEFT_LATITUDE",    -90.0)
Magics.setr ("SUBPAGE_LOWER_LEFT_LONGITUDE",  140.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LATITUDE",   -60.0)
Magics.setr ("SUBPAGE_UPPER_RIGHT_LONGITUDE",  180.0)
Magics.cont()
Magics.coast()
Magics.setc ("text_line_1", "Contours with solid shading")
Magics.seti ("text_line_count",1)
Magics.setr ("text_reference_character_height",0.5)
Magics.setc ("text_justification","centre")
Magics.text()

Magics.finalize()

File: test/fortran/shading.f
         PROGRAM CONTSOLIDSHADETEST1

C     A test of Magics++ solid shading, using lots of different areas


      PARAMETER (NLEV=11)
      DIMENSION  RLEV (NLEV)
      DATA       RLEV /48000., 49000., 50000., 51000., 52000., 
     +                  53000.,
     +                 54000., 55000., 56000., 57000., 58000./


C     Open MAGICS and set the output device

      CALL POPEN
      CALL PARSE_COMMAND_LINE ('shading')


C     Set up the coastline attributes

      CALL PSETC ('MAP_COASTLINE_COLOUR', 'GREY')
      CALL PSETC ('MAP_GRID_COLOUR',      'GREY')
      CALL PSETR ('MAP_GRID_LONGITUDE_INCREMENT', 10.0)
      CALL PSETR ('MAP_GRID_LATITUDE_INCREMENT',   5.0)
        


C     Set up the text attributes

      CALL PSETI ('TEXT_LINE_COUNT', 3)


C     Define the contour attributes

      CALL PSETC  ('LEGEND',                       'ON')
      CALL PSETC  ('CONTOUR_LEVEL_SELECTION_TYPE', 'LEVEL_LIST')
      CALL PSET1R ('CONTOUR_LEVEL_LIST',            RLEV, NLEV)      
      CALL PSETC  ('CONTOUR_LINE_COLOUR',      'BLUE')
      CALL PSETC  ('CONTOUR_HIGHLIGHT_COLOUR', 'BLUE')
      CALL PSETC  ('CONTOUR_SHADE',            'ON')      
      CALL PSETC  ('CONTOUR_SHADE_TECHNIQUE',  'POLYGON_SHADING')
      CALL PSETC  ('CONTOUR_SHADE_METHOD',     'AREA_FILL')


C     Pass the data to MAGICS

      



C     Start plotting - one page per geographical area.

C                        LEFT , RIGHT, TOP , BOTTOM, NEWPAGE?
      

      CALL PSETC ('TEXT_LINE_3', 'Bottom-right corner')
      CALL CONTOUR_TEST ( 140.,  180., -60.,  -70.,   0)


      CALL PCLOSE

      STOP
      END



C    ------------------------------------------------------------
C    CONTOUR_TEST
C    Plots solid shaded contours with coastlines within the given
C    geographical area.
C    ------------------------------------------------------------

      SUBROUTINE CONTOUR_TEST (RLEFT, RIGHT, RTOP, RBOT, INEW)
      
          CHARACTER*80  TITLE
          REAL          RLATINC, RLONGINC


C         Optionally start a new page

          IF (INEW == 1) THEN
            CALL PNEW  ('SUPER_PAGE')
          END IF

        
          

          CALL PSETR ('SUBPAGE_LOWER_LEFT_LATITUDE',    RBOT)
          CALL PSETR ('SUBPAGE_LOWER_LEFT_LONGITUDE',   RLEFT)
          CALL PSETR ('SUBPAGE_UPPER_RIGHT_LATITUDE',   RTOP)
          CALL PSETR ('SUBPAGE_UPPER_RIGHT_LONGITUDE',  RIGHT)

          CALL PSETC ('GRIB_INPUT_TYPE',      'FILE')
          CALL PSETC ('GRIB_INPUT_FILE_NAME', 'data/z500.grb')
          CALL PSETC ('GRIB_SPECIFICATION',   'OFF')
          CALL PSETC ('GRIB_SCALING',         'OFF')
          CALL PGRIB
C         Set up the geographical area

C         Calculate the best lat/long label increments

          RLATINC  = (RTOP  - RBOT)  / 10.
          RLONGINC = (RIGHT - RLEFT) / 10.

          CALL PSETR ('MAP_GRID_LONGITUDE_INCREMENT',  RLONGINC)
          CALL PSETR ('MAP_GRID_LATITUDE_INCREMENT',   RLATINC)
 



C         Draw contours and coastlines

          CALL PCONT
          CALL PCOAST


C         Plot the title

          WRITE(TITLE,'(A,4F6.1)') '(L,R,T,B): ',
     X          RLEFT, RIGHT, RTOP, RBOT
          CALL PSETC ('TEXT_LINE_2', TITLE)

          CALL PTEXT

      RETURN
      END






C --------------------------------------------------------------------
C     PARSE_COMMAND_LINE
C     Checks the command-line for any arguments.
C     Arguments can come in pairs. Currently supported arguments are:
C     PROJECTION 
C     DEVICE 
C     e.g. Run the program with:
C      PROJECTION POLAR_STEREOGRAPHIC
C      PROJECTION CYLINDRICAL   DEVICE SVG
C --------------------------------------------------------------------

      SUBROUTINE PARSE_COMMAND_LINE (OUTROOTNAME)

      CHARACTER*32 ARG
      CHARACTER*32 PROJECTION
      CHARACTER*32 DEVICE
      CHARACTER*48 EXENAME
      CHARACTER*64 OUTNAME
      CHARACTER*(*) OUTROOTNAME
      INTEGER NUM_ARGS
      INTEGER DEVICE_SET

      DEVICE_SET = 0

      NUM_ARGS = IARGC()

      I = 1

20    IF (I.LE.NUM_ARGS) THEN
          CALL GETARG ( I, ARG ) 
          

C         Set the projection?

          IF (ARG.EQ.'PROJECTION') THEN
              I = I + 1 
              CALL GETARG ( I, PROJECTION ) 
              CALL PSETC ('SUBPAGE_MAP_PROJECTION', PROJECTION)


C        Set the device?

          ELSEIF (ARG.EQ.'DEVICE') THEN
              I = I + 1 
              CALL GETARG ( I, DEVICE ) 

C             Set the output filename

              IF     (DEVICE.EQ.'PS')  THEN
                OUTNAME = OUTROOTNAME //   '.ps'
                CALL PSETC ('DEVICE',       DEVICE)
                CALL PSETC ('PS_DEVICE',   'ps_a4')
                CALL PSETC ('PS_FILE_NAME', OUTNAME)
                CALL PSETC ('PS_PDF',      'BOTH')
              ELSEIF (DEVICE.EQ.'GIF') THEN
                CALL PSETC ('DEVICE',       'GD')
                OUTNAME = OUTROOTNAME //   '.gif'
                CALL PSETC ('GD_FORMAT',   'ANIMATION')
                CALL PSETC ('GD_FILE_NAME', OUTNAME)
                CALL PSETI ('GD_DELAY',     150)
              ELSEIF (DEVICE.EQ.'PDF') THEN
                CALL PSETC ('DEVICE',       'IM')
                OUTNAME = OUTROOTNAME //   '.pdf'
                CALL PSETC ('IM_FILE_NAME', OUTNAME) 
              ELSEIF (DEVICE.EQ.'SVG') THEN
                CALL PSETC ('DEVICE',       DEVICE)
                OUTNAME = OUTROOTNAME //   '.svg'
                CALL PSETC ('SVG_FILE_NAME', OUTNAME)
              ELSE
                WRITE(*, '(A)') 'BAD DEVICE: ', DEVICE
              ENDIF
            
            DEVICE_SET = 1

          ELSEIF (ARG.EQ.'PS_SPLIT') THEN
                CALL PSETC ('PS_SPLIT',     'ON')
          ENDIF

          I = I + 1 
          GOTO 20
      ENDIF


C     If no device has been set, then use PostScript by default

      IF (DEVICE_SET.EQ.0) THEN
        OUTNAME = OUTROOTNAME // '.ps'
        CALL PSETC ('PS_DEVICE',    'ps_a4')
        CALL PSETC ('PS_FILE_NAME', OUTNAME)
      ENDIF

      END


